### PR TITLE
Add writing to CWAgent's log before a panic

### PR DESCRIPTION
**Pull request by khanhntd**
# Description of the issue
Currently, panic() does not write log message to CWAgent's log. Therefore, add a `log.Printf` before a panic based on these criteria
* As long as it does not involve in test since I believe it would be better to use exit for [short term termination](https://stackoverflow.com/a/28473339)
* Has a `sprintf` or `printf` before a panic.
* Does not involve with `internal` package since most encountered errors has been at surface levels instead of internal package and avoid having unnecessary logs.  

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Use `make test` to test the changes. After making some test with Linux, Mac and Windows, they are having the same behavior:
* Without using recover panic
image

* With using recover for panic
image

### CODE REVIEW

File: cmd/config-downloader/downloader.go [hunk 1]
```diff
@@ -125,13 +125,15 @@ func main() {
 	if inputConfig != "" {
 		f, err := os.Open(inputConfig)
 		if err != nil {
-			fmt.Printf("E! Failed to open Common Config: %v\n", err)
-			panic(err)
+			errorMessage := fmt.Sprintf("E! Failed to open Common Config: %v\n", err)
+			log.Printf(errorMessage)
```
Comment (SaxyPandaBear) [inline cmd/config-downloader/downloader.go:129]: So, two things.
1. I was reading up on the differences between `fmt.Print()` and `log.Print()` and from what I saw online, by default, the `log` package writes to stderr whereas `fmt` writes to stdout. **This would result in a change in behavior**. 
2. If we determine it's okay to accept the above change in behavior, this could be simplified to 
```suggestion
			log.Panicf("E! Failed to open Common Config: %v\n", err)
```

File: cmd/config-downloader/downloader.go [hunk 2]
```diff
@@ -154,11 +157,12 @@ func main() {
 	if region == "" && downloadLocation != locationDefault {
 		fmt.Println("Unable to determine aws-region.")
 		if mode == config.ModeEC2 {
-			errorMessage = fmt.Sprintf("Please check if you can access the metadata service. For exampe, on linux, run 'wget -q -O - http://169.254.169.254/latest/meta-data/instance-id && echo' ")
+			errorMessage = fmt.Sprintf("E! Please check if you can access the metadata service. For example, on linux, run 'wget -q -O - http://169.254.169.254/latest/meta-data/instance-id && echo' ")
```
Comment (SaxyPandaBear) [inline cmd/config-downloader/downloader.go:160]: I don't think that the downloader executable points to the telegraf logger in any way, so I don't think that the `E!` notation helps much outside of letting us know semantically that this is an "error". Just calling that out. It's not a breaking change so I don't mind keeping it.

File: cmd/config-translator/translator_test.go [hunk 3]
```diff
@@ -19,14 +19,14 @@ import (
 func checkIfSchemaValidateAsExpected(t *testing.T, jsonInputPath string, shouldSuccess bool, expectedErrorMap map[string]int) {
 	actualErrorMap := make(map[string]int)
 
-	jsonInputMap, err := util.GetJsonMapFromFile(jsonInputPath)
-	if err != nil {
-		t.Fatalf("Failed to get json map from %v with error: %v", jsonInputPath, err)
+	jsonInputMap, e := util.GetJsonMapFromFile(jsonInputPath)
+	if e != nil {
+		t.Fatalf("Failed to get json map from %v with error: %v", jsonInputPath, e)
```
Comment (SaxyPandaBear) [inline cmd/config-translator/translator_test.go:24]: This can be simplified to:
```suggestion
	assert.NoError(t, e)
```

There is a lot of code like this and I wanted to clean up our test code myself, so I'm fine with leaving this as is in your PR because I'll just update it in mine.

File: internal/containerinsightscommon/nodeCapacity.go [hunk 4]
```diff
@@ -18,7 +18,7 @@ type NodeCapacity struct {
 
 func NewNodeCapacity() *NodeCapacity {
 	if _, err := os.Lstat("/rootfs/proc"); os.IsNotExist(err) {
-		panic("/rootfs/proc doesn't exists")
+		panic("E! /rootfs/proc doesn't exists")
```
Comment (SaxyPandaBear) [inline internal/containerinsightscommon/nodeCapacity.go:21]: How is this error getting logged? This just adds the telegraf log error prefix but I don't think that this alone writes to a log about the error.

File: translator/config/schema_test.go [hunk 5]
```diff
@@ -12,9 +12,9 @@ import (
 )
 
 func TestGetJsonSchema(t *testing.T) {
-	jsonFile, err := ioutil.ReadFile("./schema.json")
-	if err != nil {
-		panic(err)
+	jsonFile, e := ioutil.ReadFile("./schema.json")
+	if e != nil {
+		panic(e)
```
Comment (SaxyPandaBear) [inline translator/config/schema_test.go:17]: This shouldn't panic, but I'm going to clean up our test code so you don't have to change it.
```suggestion
	assert.NoError(t, e)
```

File: translator/translate/metrics/append_dimensions/append_dimensions_test.go [hunk 6]
```diff
@@ -11,19 +11,19 @@ import (
 )
 
 func TestAppendDimensions(t *testing.T) {
-	e := new(appendDimensions)
+	a := new(appendDimensions)
 	//Check whether override default config
 	var input interface{}
-	err := json.Unmarshal([]byte(`{
+	e := json.Unmarshal([]byte(`{
```
Comment (SaxyPandaBear) [inline translator/translate/metrics/append_dimensions/append_dimensions_test.go:17]: This feels like an unnecessary change. Why did you change `err` to `e`? This caused you to change the variable name of the dimension.

File: cmd/config-translator/translator_test.go [hunk 7]
```diff
@@ -19,14 +19,14 @@ import (
 func checkIfSchemaValidateAsExpected(t *testing.T, jsonInputPath string, shouldSuccess bool, expectedErrorMap map[string]int) {
 	actualErrorMap := make(map[string]int)
 
-	jsonInputMap, err := util.GetJsonMapFromFile(jsonInputPath)
-	if err != nil {
-		t.Fatalf("Failed to get json map from %v with error: %v", jsonInputPath, err)
+	jsonInputMap, e := util.GetJsonMapFromFile(jsonInputPath)
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply cmd/config-translator/translator_test.go:22]: It's for differentiate between error in usual function and in testing function. However, most of others companies using golang would use `err` in their function instead. That's why I need third party opinion a little bit.

File: translator/config/schema_test.go (hunk 5 above)
Comment (khanhntd) [reply translator/config/schema_test.go:17]: Agree. Thought of doing in another PR but prefer left it as it is.

File: translator/translate/metrics/append_dimensions/append_dimensions_test.go (hunk 6 above)
Comment (khanhntd) [reply translator/translate/metrics/append_dimensions/append_dimensions_test.go:17]: It's for differentiate between error in usual function and in testing function. However, most of others companies using golang would use err in their function instead. That's why I need third party opinion a little bit.

File: cmd/config-downloader/downloader.go (hunk 2 above)
Comment (khanhntd) [reply cmd/config-downloader/downloader.go:155]: Agree on this. Just my habit of grouping things to error.

File: plugins/inputs/logfile/tail/watch/polling.go [hunk 8]
```diff
@@ -39,7 +40,9 @@ func (fw *PollingFileWatcher) BlockUntilExists(t *tomb.Tomb) error {
 			return tomb.ErrDying
 		}
 	}
-	panic("unreachable")
+	errorMessage := fmt.Sprintf("E! Unreachable to file %s with PollingFileWatcher",fw.Filename)
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply plugins/inputs/logfile/tail/watch/polling.go]: Agree. Thought deleting panics would do some breaking changes so keeping their.

[the remark this answers is not in the document]
Comment (khanhntd) [reply plugins/inputs/logfile/tail/watch/polling.go]: Agree. Thought deleting panics would do some breaking changes so keeping their.

File: internal/containerinsightscommon/nodeCapacity.go (hunk 4 above)
Comment (khanhntd) [reply internal/containerinsightscommon/nodeCapacity.go:21]: That is one of the criteria I mentioned: `most encountered errors has been at surface levels instead of internal package and avoid having unnecessary logs.`. However, if it is needed, I would add logs for this as well.

File: plugins/inputs/logfile/tail/tail.go [hunk 9]
```diff
@@ -593,7 +596,7 @@ func (tail *Tail) exitOnDeletion() {
 // with the last chunk of variable size.
 func partitionString(s string, chunkSize int) []string {
 	if chunkSize <= 0 {
-		panic("invalid chunkSize")
+		panic("Invalid chunkSize")
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply plugins/inputs/logfile/tail/tail.go:599]: It does not write log before panicking. Same criteria that I mentioned before: `most encountered errors has been at surface levels instead of internal package and avoid having unnecessary logs`. However, if it is needed, I would add writing to logs for this as well

File: plugins/inputs/logfile/tail/tail.go [hunk 10]
```diff
@@ -465,7 +465,10 @@ func (tail *Tail) waitForChanges() error {
 	case <-tail.Dying():
 		return ErrStop
 	}
-	panic("unreachable")
+	errorMessage := fmt.Sprintf("E! Unreachable to file %s",tail.Filename)
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply plugins/inputs/logfile/tail/tail.go:468]: Agree. Thought deleting panics would do some breaking changes so keeping their.

File: cmd/config-translator/translator_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply cmd/config-translator/translator_test.go:22]: I don't understand what the concern of having it referred to as `err` is. This is in a test so there's no reason anyone other than us would use this.

[the remark this answers is not in the document]
Comment (khanhntd) [reply cmd/config-translator/translator_test.go:22]: I'm good with either so it should not be hard to change it back.

File: cfg/aws/refreshable_shared_credentials_provider.go [hunk 11]
```diff
@@ -22,7 +22,7 @@ type Refreshable_shared_credentials_provider struct {
 func (p *Refreshable_shared_credentials_provider) Retrieve() (credentials.Value, error) {
 
 	p.SetExpiration(time.Now().Add(p.ExpiryWindow), 0)
-	creds, e := p.sharedCredentialsProvider.Retrieve()
+	creds, err := p.sharedCredentialsProvider.Retrieve()
```
Comment (SaxyPandaBear) [inline cfg/aws/refreshable_shared_credentials_provider.go:25]: I'm okay with this, but if we're going to be updating a lot of variable names just because we want to change the name of the variables, it would help if that was in a separate PR. This PR already has a lot of changes because of all of the panic logging we are adding. Including cosmetic changes to unrelated source code for variable names makes this harder to review

File: cmd/config-downloader/downloader.go [hunk 12]
```diff
@@ -125,26 +125,24 @@ func main() {
 	if inputConfig != "" {
 		f, err := os.Open(inputConfig)
 		if err != nil {
-			fmt.Printf("E! Failed to open Common Config: %v\n", err)
-			panic(err)
+			log.Panicf("E! Failed to open Common Config: %v", err)
```
Comment (SaxyPandaBear) [inline cmd/config-downloader/downloader.go:128]: Did you confirm on all major OS types that these logs are visible in the agent log even though they are going to stderr instead of stdout?

File: cmd/config-downloader/downloader.go [hunk 13]
```diff
@@ -187,7 +185,8 @@ func main() {
 
 	locationArray := strings.SplitN(downloadLocation, locationSeparator, 2)
 	if locationArray == nil || len(locationArray) < 2 && downloadLocation != locationDefault {
-		panic(fmt.Sprintf("downloadLocation %s is malformated.\n", downloadLocation))
+		errorMessage := fmt.Sprintf("E! downloadLocation %s is malformated.", downloadLocation)
+		log.Panicf(errorMessage)
```
Comment (SaxyPandaBear) [inline cmd/config-downloader/downloader.go:189]: If we're using `Panicf`, we don't need to use `fmt.Sprintf` to create the error message beforehand.

File: cmd/config-translator/translator.go [hunk 14]
```diff
@@ -90,16 +89,16 @@ func main() {
...
 	if !ctx.RunInContainer() {
 		// run as user only applies to non container situation.
-		current, e := user.Current()
-		if e == nil && current.Name == "root" {
+		current, err := user.Current()
+		if err == nil && current.Name == "root" {
 			runAsUser, err := cmdutil.DetectRunAsUser(mergedJsonConfigMap)
 			if err != nil {
-				panic("E! Failed to detectRunAsUser\n")
+				log.Panicf("E! Failed to detectRunAsUser")
```
Comment (SaxyPandaBear) [inline cmd/config-translator/translator.go:101]: If we aren't passing in any other parameters for the format string, this could just be:
```suggestion
				log.Panic("E! Failed to detectRunAsUser")
```

File: plugins/inputs/logfile/tail/tail.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (SaxyPandaBear) [reply plugins/inputs/logfile/tail/tail.go:596]: Why do we not want to write any log for this if it panics? Are you saying that telegraf is catching this before terminating?

File: plugins/inputs/windows_event_log/wineventlog/utils_test.go [hunk 15]
```diff
@@ -46,15 +46,15 @@ func TestDecodingWithDifferentBufferUsedReturned(t *testing.T) {
...
 	data, _ := hex.DecodeString(originalHexStr)
-	bufferUsed := len(data)/2
+	bufferUsed := len(data) / 2
```
Comment (SaxyPandaBear) [inline plugins/inputs/windows_event_log/wineventlog/utils_test.go:49]: Why is go fmt inconsistent when doing operations? I noticed that when formatting, it removes spaces around `+` but here, it added spaces around `/`

File: translator/cmdutil/translatorutil.go [hunk 16]
```diff
@@ -101,16 +101,16 @@ func RunSchemaValidation(inputJsonMap map[string]interface{}) (*gojsonschema.Res
 func checkSchema(inputJsonMap map[string]interface{}) {
 	result, err := RunSchemaValidation(inputJsonMap)
 	if err != nil {
-		panic(err.Error())
+		log.Panicf("E! Failed to run schema validation because of %v", err)
 	}
 	if result.Valid() {
-		fmt.Println("Valid Json input schema.")
+		log.Print("I! Valid Json input schema.")
```
Comment (SaxyPandaBear) [inline translator/cmdutil/translatorutil.go:107]: Why did this get changed to a log? This introduces more inconsistencies. We use `fmt.Print` for a bunch of things to write to stdout for normal success paths, but this changes this specific function to still log to stderr

File: translator/config/os.go [hunk 17]
```diff
@@ -22,11 +22,14 @@ func ToValidOs(os string) string {
...
 	formattedOs := strings.ToLower(os)
 	for _, val := range supportedOs {
 		if formattedOs == val {
 			return formattedOs
 		}
 	}
-	panic(fmt.Sprintf("%v is not a supported osType type", os))
+
+	log.Panicf("E! %v is not a supported osType type", os)
+	return ""
```
Comment (SaxyPandaBear) [inline translator/config/os.go:34]: This is unreachable. Frankly, I dislike this because the logic here is fragile. I think that this function should not panic at all. It should either:
1. return an empty string and an error if the input is not a valid OS
2. just return an empty string and code that uses it needs to know to check for the empty string when using the return value from this function.

If that's too much to add to this PR, I would revert the changes to this function and address it separately.

File: translator/totomlconfig/toTomlConfig.go [hunk 18]
```diff
@@ -58,9 +58,9 @@ func ToTomlConfig(c interface{}) string {
 	_, val := r.ApplyRule(c)
 	buf := bytes.Buffer{}
 	enc := toml.NewEncoder(&buf)
-	e := enc.Encode(val)
-	if e != nil {
-		panic(e)
+	err := enc.Encode(val)
+	if err != nil {
+		panic(err)
```
Comment (SaxyPandaBear) [inline translator/totomlconfig/toTomlConfig.go:63]: Why don't we log this err?

File: translator/translate/logs/logs_collected/windows_events/collect_list/collectlist_test.go [hunk 19]
```diff
@@ -59,11 +59,11 @@ func TestApplyRule(t *testing.T) {
 
 	var actual interface{}
 
-	error := json.Unmarshal([]byte(rawJsonString), &input)
-	if error == nil {
+	e := json.Unmarshal([]byte(rawJsonString), &input)
```
Comment (SaxyPandaBear) [inline translator/translate/logs/logs_collected/windows_events/collect_list/collectlist_test.go:62]: This is not consistent with all of the other name changing. Why is this not `err`?

File: translator/translate/metrics/append_dimensions/append_dimensions_test.go [hunk 20]
```diff
@@ -11,7 +11,7 @@ import (
 )
 
 func TestAppendDimensions(t *testing.T) {
-	e := new(appendDimensions)
+	a := new(appendDimensions)
```
Comment (SaxyPandaBear) [inline translator/translate/metrics/append_dimensions/append_dimensions_test.go:14]: The changes to this file can be reverted.

File: translator/translate/metrics/drop_origin/drop_origin.go [hunk 21]
```diff
@@ -36,7 +36,7 @@ func (do *dropOrigin) ApplyRule(input interface{}) (returnKey string, returnVal
 						if reflect.TypeOf(value).String() == "string" {
 							droppingDimensions = append(droppingDimensions, value.(string))
 						} else {
-							log.Panic("Fail to translate the JSON config, invalid format of dropping dimensions.")
+							log.Panicf("Fail to translate the JSON config, invalid format of dropping dimensions.")
```
Comment (SaxyPandaBear) [inline translator/translate/metrics/drop_origin/drop_origin.go:39]: Why does this need `Panicf`?

File: translator/translate/metrics/metrics_collect/gpu/nvidiaSmi_test.go [hunk 22]
```diff
@@ -43,7 +43,7 @@ func TestNoFieldConfig(t *testing.T) {
 func TestFullConfig(t *testing.T) {
 	n := new(NvidiaSmi)
 	var input interface{}
-	err := json.Unmarshal([]byte(`{"nvidia_gpu":{"measurement": [
+	e := json.Unmarshal([]byte(`{"nvidia_gpu":{"measurement": [
```
Comment (SaxyPandaBear) [inline translator/translate/metrics/metrics_collect/gpu/nvidiaSmi_test.go:46]: Why is this changed to `e`? It's not consistent.

File: translator/translate/metrics/rollup_dimensions/rollupDimensions_test.go [hunk 23]
```diff
@@ -13,13 +13,13 @@ import (
 )
 
 func TestRollupDimensions(t *testing.T) {
-	e := new(rollupDimensions)
+	r := new(rollupDimensions)
```
Comment (SaxyPandaBear) [inline translator/translate/metrics/rollup_dimensions/rollupDimensions_test.go:16]: This change can be reverted

**Comment by SaxyPandaBear** [comment]
Also it would be great to share any manual testing results with the built agent executables with this new change in behavior to use the `log` package instead of `fmt` for writing out our logging. I am a little concerned about what will happen with where the logs go. I think a requirement before approving is to take one of the code paths that leads to a `log.Panic` call and reproduce it on Darwin, Linux and Windows in order to confirm that the log shows up in the same place as where it used to before your changes

File: cmd/config-translator/translator.go (hunk 14 above)
Comment (khanhntd) [reply cmd/config-translator/translator.go:101]: Have not notice that. I used `log.Panic` for some but not noticing this until now. Nice catch

File: translator/translate/logs/logs_collected/windows_events/collect_list/collectlist_test.go (hunk 19 above)
Comment (khanhntd) [reply translator/translate/logs/logs_collected/windows_events/collect_list/collectlist_test.go]: Fixed it in the newest commit. Nice catch on these though.

File: translator/translate/metrics/append_dimensions/append_dimensions_test.go (hunk 20 above)
Comment (khanhntd) [reply translator/translate/metrics/append_dimensions/append_dimensions_test.go]: Also fixed it in the newest commit. Another nice catch on these though.

File: translator/config/os.go (hunk 17 above)
Comment (khanhntd) [reply translator/config/os.go]: Agree on fixing it in a separate PR and fixed it in the newest commit.

File: translator/translate/agent/ruleLogFile.go [hunk 24]
```diff
@@ -33,7 +33,8 @@ func getDefaultValue() string {
 	case config.OS_TYPE_WINDOWS:
 		return util.GetWindowsProgramDataPath() + "\\Amazon\\AmazonCloudWatchAgent\\Logs\\amazon-cloudwatch-agent.log"
 	default:
-		panic(fmt.Sprintf("Unsupported platform %v for logRule", targetPlatform))
+		log.Panicf("Unsupported platform %v for logRule", targetPlatform)
+		return ""
```
[the remark this answers is not in the document]
Comment (khanhntd) [reply translator/translate/agent/ruleLogFile.go]: Agree on the aforementioned solution and will fix it in a separate PR and fixed it in the newest commit.

File: cmd/config-downloader/downloader.go (hunk 13 above)
Comment (khanhntd) [reply cmd/config-downloader/downloader.go]: Fixed it in the newest commit. Nice catch on these though.

File: translator/totomlconfig/toTomlConfig.go (hunk 18 above)
Comment (khanhntd) [reply translator/totomlconfig/toTomlConfig.go]: Cause I thought [this log](https://github.com/aws/amazon-cloudwatch-agent/blob/master/translator/cmdutil/translatorutil.go#L45-L47) would be enough for us to track down  the issue. However, it could be from `BurntSushi` itself so reconsider to add the error.

File: translator/cmdutil/translatorutil.go (hunk 16 above)
Comment (khanhntd) [reply translator/cmdutil/translatorutil.go:107]: Mostly changing this would signal us that CWAgent translated successfully and moving on to other executables in the log. I am considering changing most of the fmt to `log.Print` since some could help us in tracking down the issues though.

File: cmd/config-downloader/downloader.go (hunk 12 above)
Comment (SaxyPandaBear) [reply cmd/config-downloader/downloader.go:128]: Just want to confirm this before approving

**Review by SaxyPandaBear — APPROVED** [review]
